tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -0,0 +1,201 @@
+    size_t size;
+    __CPROVER_assume(size <= messageLength);
+    return size;
+}
+
+IotNetworkError_t IotNetworkInterfaceClose( void * pConnection ) {
+    IotNetworkError_t error;
+    return error;
+}
+
+size_t IotNetworkInterfaceReceive( void * pConnection,
+        uint8_t * pBuffer,
+        size_t bytesRequested ) {
+    uint8_t byte;
+    __CPROVER_array_copy(pBuffer,&byte);
+
+    size_t size;
+    __CPROVER_assume(size <= bytesRequested);
+    return size;
+}
+
+IotNetworkError_t IotNetworkInterfaceCallback( void * pConnection,
+                 IotNetworkReceiveCallback_t
+                 receiveCallback,
+                 void * pContext ) {
+    IotNetworkError_t error;
+    return error;
+}
+
+IotNetworkError_t IotNetworkInterfaceDestroy( void * pConnection ) {
+  IotNetworkError_t error;
+  return error;
+}
+
+IotNetworkInterface_t IOTNI = {
+  .create = IotNetworkInterfaceCreate,
+  .close = IotNetworkInterfaceClose,
+  .send = IotNetworkInterfaceSend,
+  .receive = IotNetworkInterfaceReceive,
+  .setReceiveCallback = IotNetworkInterfaceCallback,
+  .destroy = IotNetworkInterfaceDestroy
+};
+
+/* Models the Network Interface. */
+IotNetworkInterface_t *newNetworkInterface() {
+  return &IOTNI;
+}
+
+/* Creates a Connection Info and assigns memory accordingly. */
+IotHttpsConnectionInfo_t * newConnectionInfo() {
+  IotHttpsConnectionInfo_t * pConnInfo = safeMalloc(sizeof(IotHttpsConnectionInfo_t));
+  if(pConnInfo) {
+    pConnInfo->pNetworkInterface = newNetworkInterface();
+    /* The +1 is required for coverage */
+    __CPROVER_assume(pConnInfo->addressLen >= 0 && pConnInfo->addressLen <= IOT_HTTPS_MAX_HOST_NAME_LENGTH + 1);
+    pConnInfo->pAddress = safeMalloc(pConnInfo->addressLen);
+    /* The +1 is required for coverage */
+    __CPROVER_assume(pConnInfo->alpnProtocolsLen >= 0 && pConnInfo->alpnProtocolsLen <= IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH + 1);
+    pConnInfo->pAlpnProtocols = safeMalloc(pConnInfo->alpnProtocolsLen);
+    pConnInfo->pCaCert = malloc(sizeof(uint32_t));
+    pConnInfo->pClientCert = malloc(sizeof(uint32_t));
+    pConnInfo->pPrivateKey = malloc(sizeof(uint32_t));
+    pConnInfo->userBuffer.pBuffer = safeMalloc(sizeof(_connHandle_t));
+  }
+  return pConnInfo;
+}
+
+/* Creates a Connection Handle and assigns memory accordingly. */
+IotHttpsConnectionHandle_t newIotConnectionHandle () {
+  IotHttpsConnectionHandle_t pConnectionHandle = safeMalloc(sizeof(_connHandle_t));
+  if(pConnectionHandle) {
+    pConnectionHandle->pNetworkConnection = safeMalloc(sizeof(uint32_t));
+    pConnectionHandle->pNetworkInterface = newNetworkInterface();
+    pConnectionHandle->reqQ.pPrevious = safeMalloc(sizeof(IotLink_t));
+    pConnectionHandle->reqQ.pNext = safeMalloc(sizeof(IotLink_t));
+    pConnectionHandle->taskPoolJob = malloc(sizeof(struct _taskPoolJob));
+  }
+  return pConnectionHandle;
+}
+
+/* Creates a Response Handle and assigns memory accordingly. */
+IotHttpsResponseHandle_t newIotResponseHandle() {
```

Comment:
Could we confirm with Sarena that these constructors really are generating unconstrained values for handles?  Is it expected that we have internal pointers like pResponseHandle->pHttpsRequest->pHttpsResponse = pResponseHandle?  Is it always the case that a current pointer points to a single thing as in pResponseHandle->pHeadersCur = pResponseHandle->pHeaders?  Since we are trying to prove memory safety, a response with a body consisting of a single word raises red flags without explanations.

Even if the answer to a question like this is "no", the work still adds value and should be added to the repository asap.  But we need to document any shortcomings.

Reply:
I changed the size of pResponseHandle->pBody to uint32_t. @sarenameas said this size is  arbitrary.

Reply:
The application can set any body size. Servers do send HTTP bodies consisting of a single word. The headers will not e a single word, so that has sizes that she has defined at the top of this file.